crates/ra_hir/src/generics.rs

```diff
@@ -87,4 +87,17 @@ impl GenericParams {
         let parent_count = self.count_parent_params();
         parent_count + self.params.len()
     }
+
+    fn for_each_param<'a>(&'a self, f: &mut impl FnMut(&'a GenericParam)) {
```

Comment:
This can't be an impl Iterator because the type is recursive, right? I wonder if we need some kind of "InternalIterator" helper crate for such cases? I imagine API like this:

```rust
pub fn params_including_parent(&'a self) -> impl InternalIterator<&'a GenericParam> + 'a {
    internal_iter(|f| {
         ....
    })
}
```

I feel it might run into GADs though...

Reply:
I've  wondere if we should make type strictly two-level. I think currently we could have at most two-level nesting, right? But I imagine extensions to the language which allow arbitrary nesting, so handling the general case seems better.

Reply:
Yes, the recursion is the problem... I'm hoping we'll be able to handle something like this with generators someday :smile: